### DISCUSSION TITLE
Optimal way to proxy large files through axum using reqwest?

### DISCUSSION THREAD

**Discussion by bes** [tokio-rs/axum#1821] [Q&A] [answered]
I'm trying to proxy large files through `axum`, currently trying to figure out how to do that using async `Client::get` from `reqwest`, but not finding anything online that might help me.

What I have already tried:

```rust
async fn proxy_req(/* parameters */) -> impl IntoResponse {
    let reqwest_response = client.get(&uri).send.await.unwrap();
    // Here I can get bytes from the response and return that
    let data = reqwest_response.bytes().await.unwrap();
    data

    // But how do I convert `reqwest_response` into a stream of data that
    // doesn't blow through all of my RAM if the response is huge, and is done
    // in a streaming / async fashion?
}
```

I am open to trying other HTTP clients if that helps.

Thanks 🙏

**Comment by davidpdrsn** [comment, marked as answer]
Call [`Response::bytes_stream`](https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.bytes_stream) and do something like this https://github.com/tokio-rs/axum/blob/main/examples/stream-to-file/src/main.rs#L104

**Comment by bes** [comment]
Thanks @davidpdrsn, your answer pointed me in the right direction! Using [Response::bytes_stream](https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.bytes_stream) and [axum::body::StreamBody](https://docs.rs/axum/latest/axum/body/struct.StreamBody.html) seems to do the right thing.

```rust
async fn proxy_req(/* parameters */) -> impl IntoResponse {
    let reqwest_response = client.get(&uri).send.await.unwrap();
    let stream = reqwest_response.bytes_stream();
    StreamBody::new(stream)
}
```
